Test reading sharded BAM files

tomwhite opened the pull request:
Fixes #935 (and #977).

tomwhite commented:
Ready for review now - @droazen, would you be able to take a look?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java:
```diff
@@ -110,13 +113,26 @@ public ReadsSparkSource(JavaSparkContext ctx) {
+                    }
+                });
+                if (bamFiles.length == 0) {
+                    throw new GATKException("No BAM files to load header from in: " + path);
```
This should probably be a UserException rather than a GATKException since it's almost certainly from someone specifying a bad path.

tomwhite replied:
Done

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java:
```diff
@@ -26,22 +26,22 @@
     @Test(dataProvider = "loadReadsBAM", groups = "spark")
     public void readsSinkTest(String inputBam, String outputFile) throws IOException {
-        new File(outputFile).deleteOnExit();
+        new File(outputFile).delete();
```
Would `IOUtils.deleteRecursivelyOnExit()` help here?  It would be best if the test doesn't leave any output when it's done.

tomwhite replied:
Agreed. Should existing output be overwritten, or is it an error in GATK? (In Hadoop/Spark it's generally an error.)

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java:
```diff
@@ -64,9 +64,26 @@ public void readsSinkTest(String inputBam, String outputFile) throws IOException
+
+        ReadsSparkSource readSource = new ReadsSparkSource(ctx);
+        JavaRDD<GATKRead> rddParallelReads = readSource.getParallelReads(inputBam);
+        rddParallelReads = rddParallelReads.repartition(2);
```
Is this guaranteeing that the output is in 2 shards?  Maybe add a comment here?  An explicit check once it's written to disk that it actually wrote 2 files would be good since I can imagine HadoopBAM could theoretically coalesce internally when the file sie is small.

tomwhite replied:
Repartition will create the given number of partitions. I'll add a check for the number of files though.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java:
```diff
@@ -110,13 +113,26 @@ public ReadsSparkSource(JavaSparkContext ctx) {
+            }
+            return SAMHeaderReader.readSAMHeaderFrom(path, ctx.hadoopConfiguration());
         } catch (IOException e) {
             throw new GATKException("unable to loader header: " + e);
```
probably also a user exception here

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java:
```diff
@@ -110,13 +113,26 @@ public ReadsSparkSource(JavaSparkContext ctx) {
+                    private static final long serialVersionUID = 1L;
+                    @Override
+                    public boolean accept(Path path) {
+                        return path.getName().startsWith("part-");
```
Is this prefix alway guaranteed or is it configurable somewhere?  We should either use the variable where this is defined directly or extract it as a constant.  (I have no idea where it's defined, by maybe you do?)

tomwhite replied:
It's a Hadoop thing, so I'll extract as a constant.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java:
```diff
@@ -64,9 +64,26 @@ public void readsSinkTest(String inputBam, String outputFile) throws IOException
 
+    @Test(dataProvider = "loadReadsBAM", groups = "spark")
+    public void readsSinkShardedTest(String inputBam, String outputFile) throws IOException {
+        new File(outputFile).delete();
```
`IOUtils.deleteRecursivelyOnExit` here too I think. We should probably add a `BaseTest.createTmpDir` that includes an automatic marking for delete at some point.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java:
```diff
@@ -64,9 +64,26 @@ public void readsSinkTest(String inputBam, String outputFile) throws IOException
     @Test(dataProvider = "loadReadsADAM", groups = "spark")
     public void readsSinkADAMTest(String inputBam, String outputFile) throws IOException {
-        new File(outputFile).deleteOnExit();
+        new File(outputFile).delete();
```
This should also delete on shutdown

lbergelson commented:
@tomwhite I had a few comments, but I take it that it pretty much "just works"? That's good news!

davidadamsphd commented:
Sorry to jump in here late, but... I'd like to see a (small) sharded BAM checked into the repo and have _that_ used for the test. @lbergelson, is that OK with you?

tomwhite commented:
I've addressed the feedback so far - except for @davidadamsphd's last comment about checking in a sharded BAM. @lbergelson let me know if you'd like me to do that; also, what do you think about the overwriting behaviour?

akiezun commented:
yes, the test is for shared bams, so we should have one checked in. It can+should be small.

tomwhite commented:
Thanks @akiezun. I've added a small sharded BAM file and a test to read it.

davidadamsphd commented:
Ping. Is this on Adam's plate or Tom's?
IMO, it looks good to merge.

akiezun commented:
it was @tomwhite's  
Merge away then

droazen commented:
@tomwhite Could you rebase this branch and merge once tests pass?
